### PR TITLE
Smart contracts 101: counter contract

### PR DESCRIPTION
**Pull request by dg-concordium**
## Purpose

Tutorial for a very simple counter contract.

## Changes

New tutorial with setup based on previous tutorials and SC for counter.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

_Remove if not applicable.

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [x] I accept the above linked CLA.

### CODE REVIEW

File: source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+        let param: IncrementVal = ctx.parameter_cursor().get()?;
+        let state = host.state_mut();
+        ensure!(
+            ctx.sender().matches_account(&ctx.owner()),
+            Error::OwnerError
+        );
+
+        ensure!(param < 0, Error::DecrementError);
+        state.counter -= param;
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst:132]: So you ensure it is negative, but then you subtract it. Is that desired? Calling it `decrement` in that case seems misleading.

File: source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        Ok(())
+    }
+
+View function
+-------------
+
+The view function will return only the counters value so you need to update its return value as u8 and return it from the host.state().
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst:139]:
```suggestion
The view function will return only the counters value so you need to update its return value as i8 and return it from the host.state().
```

File: source/mainnet/smart-contracts/tutorials/counter/index.rst [hunk 3]
```diff
@@ -0,0 +1,14 @@
+.. _counter-sc-index:
+
+=================================
+Concordium counter smart contract
+=================================
+
+This tutorial guides you through creating a smart contract using the Concordium a default contract template that simply keeps a counter value in its state. It is a super simple, fundamental example contract that touches on the following points: to be able to increase/decrease the counter value by the parameter given by the user if it is a positive number, view the current value, return a custom error when someone tries to increase it with a negative value (or vice versa), and all these operations have to be done by only the owner of the contract.
+
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/index.rst:8]: I like this.

File: source/mainnet/smart-contracts/tutorials/counter/setup-env.rst [hunk 4]
```diff
@@ -0,0 +1,101 @@
+.. include:: ../../../variables.rst
+.. _setup-env:
+
+=================================
+Setup the development environment
+=================================
+
+First, you need to install “rustup” which installs Rust and Cargo to your computer. Go to `Rustup <https://rustup.rs/>`_ to install ``rustup`` for your platform.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/setup-env.rst:8]:
```suggestion

First, you need to install ``rustup`` which installs Rust and Cargo to your computer. Go to `Rustup <https://rustup.rs/>`_ to install ``rustup`` for your platform.
```

File: source/mainnet/smart-contracts/tutorials/counter/setup-env.rst [hunk 5]
```diff
@@ -0,0 +1,101 @@
+.. include:: ../../../variables.rst
+.. _setup-env:
+
+=================================
+Setup the development environment
+=================================
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/setup-env.rst:6]: Why are we repeating this? Don't we already have the exact same instructions elsewhere where we could link to them?

File: source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst (hunk 1 above)
Comment (bogacyigitbasi) [reply source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst:138]: Line 132 should be "state.counter += param; "

File: source/mainnet/smart-contracts/tutorials/counter/setup-env.rst (hunk 5 above)
Comment (dg-concordium) [reply source/mainnet/smart-contracts/tutorials/counter/setup-env.rst]: So I have completed this, also for the Piggy Bank. All tutorials now point to the one setup dev environment topic.

File: source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+-----------------
+
+Then change the update function as described below, let’s remember what we want; input needs to be parsed without any errors, the value should be positive(otherwise we will get an Error::IncrementError and the transaction should be triggered by the owner of the contract instance or it will throw Error::OwnerError and the function itself has to be a mutable function because we are going to change the state of the contract.
+
+.. code-block:: rust
+
+    type IncrementVal = i8;
+    /// Receive function. The input parameter is the boolean variable `throw_error`.
+    ///  If `throw_error == true`, the receive function will throw a custom error.
+    ///  If `throw_error == false`, the receive function executes successfully.
```
[the remark this answers is not in the document]
Comment (dg-concordium) [reply source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst:87]: @bogacyigitbasi What was the corrected text so it matches what you have in the example contract?

File: source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+Increment counter
+-----------------
+
+Then change the update function as described below. Remember that input needs to be parsed without any errors. The value must be positive, otherwise you will get an ``Error::IncrementError``. The transaction must be triggered by the owner of the contract instance or it will throw ``Error::OwnerError``. And the function itself has to be a mutable function because you are going to change the state of the contract.
+
+.. code-block:: rust
+
+    type IncrementVal = i8;
+    /// Receive function. The input parameter is the increment value `i8`.
+    ///  If the account owner does not match the contract owner, the receive function will throw a custom error.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst:86]:
```suggestion
    ///  If the account owner does not match the contract owner, the receive function will throw [`Error::OwnerError`].
```

File: source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    ///  If the number to increment by is negative, the receive function will throw a custom error.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst:87]:
```suggestion
    ///  If the number to increment by is not positive, the receive function will throw [`Error::IncrementError`].
```
(0 is also not allowed).

File: source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        _ctx: &'a impl HasReceiveContext,
+        host: &'b impl HasHost<State, StateApiType = S>,
+    ) -> ReceiveResult<i8> {
+        Ok(host.state().counter)
+    }
+
+Build, deploy, and initialize the contract
+==========================================
+
+Create a dist folder to keep the schema output file and Wasm compiled contract in and run the build command.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst:164]:
```suggestion
Create a ``dist`` folder to keep the schema output file and Wasm compiled contract in and run the build command.
```

File: source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+.. image:: images/owner-error-ok.png
+    :width: 100%
+
+Now check the state once more.
+
+.. image:: images/invoke2.png
+    :width: 100%
+
+Unsurprisingly, the state is 2. Now check the other requirement: that you cannot increment it with a negative number. Change the value in the json file to a negative number like -2.
```
Comment (abizjak) [inline source/mainnet/smart-contracts/tutorials/counter/counter-contract.rst:232]:
```suggestion
Unsurprisingly, the state is 2. Now check the other requirement: that you cannot increment it with a negative number. Change the value in the JSON file to a negative number like -2.
```

**Review by abizjak — APPROVED** [review]

**Pull request merged**